### DISCUSSION TITLE
How to shift data augmentations to GPU

### DISCUSSION THREAD

**Discussion by rdilip** [google/flax#2724] [Q&A] [answered]
A common pattern is to use the tf.Data api with Flax neural networks. Sometimes there are additional data augmentations that would benefit from having GPU access. However, I believe Jax and TF will both try to pre-allocate 90% of the GPU, so you need to turn off Tensorflow's access to the GPU. How should I take advantage of the GPU for data preprocessing? Should I just do data loading on the CPU, then do augmentations in Jax so that can take place on the GPU?

**Comment by cgarciae** [comment, marked as answer]
There are many ways to solve this.

* [Controlling memory allocation for JAX](https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html), you could try using e.g. `XLA_PYTHON_CLIENT_MEM_FRACTION=.80` (or some other fraction), or some of the other options.
* Preprocess each batch with JAX before passing it to the `train_step`.

All that said, I'd be surprised if your training is bound by the data augmentation process or the data pipeline in general. One thing you can try if this is the case is running the data augmentation process once and saving the results as a new dataset and do training on that.
